File: Backend/Microservices/PostService/PostService.Domain/Entities/PostInteraction.cs
using MongoDB.Bson;
using MongoDB.Bson.Serialization.Attributes;

namespace PostService.Domain.Entities;

public class PostInteraction
{
  [BsonId]
  [BsonRepresentation(BsonType.ObjectId)]
  public string Id { get; set; } = ObjectId.GenerateNewId().ToString();

  [BsonElement("SubjectId")]
  public string SubjectId { get; set; }

  [BsonElement("PostId")]
  public string PostId { get; set; }

  [BsonElement("Amazed")]
  public bool Amazed { get; set; }

  [BsonElement("Celebration")]
  public bool Celebration { get; set; }

  [BsonElement("ReachedTarget")]
  public bool ReachedTarget { get; set; }

  [BsonElement("Flame")]
  public bool Flame { get; set; }

  [BsonElement("LostMind")]
  public bool LostMind { get; set; }
}

File: MobileApp/gymly/lib/models/post_interaction.dart
class PostInteraction {
  String id;
  final String subjectId;
  final String postId;
  final bool amazed;
  final bool celebration;
  final bool reachedTarget;
  final bool flame;
  final bool lostMind;

  PostInteraction(
    this.id,
    this.subjectId,
    this.postId,
    this.amazed,
    this.celebration,
    this.reachedTarget,
    this.flame,
    this.lostMind,
  );

  factory PostInteraction.fromJson(Map<String, dynamic> json) {
    return PostInteraction(
      json['id'] as String,
      json["subjectId"] as String,
      json["postId"] as String,
      json["amazed"] as bool,
      json["celebration"] as bool,
      json["reachedTarget"] as bool,
      json["flame"] as bool,
      json["lostMind"] as bool,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      "id": id,
      "subjectId": subjectId,
      "postId": postId,
      "amazed": amazed,
      "celebration": celebration,
      "reachedTarget": reachedTarget,
      "flame": flame,
      "lostMind": lostMind,
    };
  }

  PostInteraction copyWith({
    String? id,
    String? subjectId,
    String? postId,
    bool? amazed,
    bool? celebration,
    bool? reachedTarget,
    bool? flame,
    bool? lostMind,
  }) {
    return PostInteraction(
      id ?? this.id,
      subjectId ?? this.subjectId,
      postId ?? this.postId,
      amazed ?? this.amazed,
      celebration ?? this.celebration,
      reachedTarget ?? this.reachedTarget,
      flame ?? this.flame,
      lostMind ?? this.lostMind,
    );
  }
}
